src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -49,19 +49,33 @@ export class DeviceSource<T extends DeviceType> {
 export class DeviceSourceManager implements IDisposable {
     // Public Members
     /**
-     * Observable to be triggered when before a device is connected
+     * Observable to be triggered when before a device is connected, any new observers added will be triggered against already connected devices
      */
-    public readonly onBeforeDeviceConnectedObservable = new Observable<{ deviceType: DeviceType, deviceSlot: number }>();
+    public readonly onBeforeDeviceConnectedObservable = new Observable<{ deviceType: DeviceType, deviceSlot: number }>((observer) => {
+        this.getListofDevices().forEach((device) => {
+            const deviceType: DeviceType = device.deviceType;
+            const deviceSlot: number = device.deviceSlot;
+
+            this.onBeforeDeviceConnectedObservable.notifyObserver(observer, { deviceType, deviceSlot });

```

Comment:
There is a small ordering issue with this. If someone adds to onAfterDeviceConnected before onBeforeDeviceConnected, then it will fire onAfterDeviceConnected first. Not sure if this will be an issue though.

I'm still not exactly sure why we have on before/after device connected events.

Reply:
Removing observables as discussed offline